<a href="https://colab.research.google.com/github/ogut77/BA/blob/main/data_manipulation_with_pandasans.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Transforming DataFrames

<p class="chapter__description">
    Let’s master the pandas basics. Learn how to inspect DataFrames and perform fundamental manipulations, including sorting rows, subsetting, and adding new columns.
  </p>
  


## Introducing DataFrames





### Inspecting a DataFrame


<div class>
<p>When you get a new DataFrame to work with, the first thing you need to do is explore it and see what it contains. There are several useful methods and attributes for this.</p>
<ul>
<li>
<code>.head()</code> returns the first few rows (the “head” of the DataFrame).</li>
<li>
<code>.info()</code> shows information on each of the columns, such as the data type and number of missing values.</li>
<li>
<code>.shape</code> returns the number of rows and columns of the DataFrame.</li>
<li>
<code>.describe()</code> calculates a few summary statistics for each column.</li>
</ul>
<p><code>homelessness</code> is a DataFrame containing estimates of homelessness in each U.S. state in 2018. The <code>individual</code> column is the number of homeless individuals not part of a family with children. The <code>family_members</code> column is the number of homeless individuals part of a family with children. The <code>state_pop</code> column is the state's total population.</p>
<p><code>pandas</code> is imported for you.</p>
</div>


In [ ]:

import pandas as pd
hs = pd.read_csv('https://raw.githubusercontent.com/ogut77/BA/main/homelessness.csv',index_col=0)



<li>Print the head of the <code>homelessness</code> DataFrame.</li>


In [ ]:
# Print the head of the homelessness data
hs.head()


region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588

In [ ]:
#Load when data is in your pc or  laptop.
hs2=pd.read_csv('/content/homelessness.csv',index_col=0)
hs2.head()

region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588

In [ ]:
#Print first two rows of homelessness data
hs2.head(2)

region    state  individuals  family_members  state_pop
0  East South Central  Alabama       2570.0           864.0    4887681
1             Pacific   Alaska       1434.0           582.0     735139


<li>Print information about the column types and missing values in <code>homelessness</code>.</li>


In [ ]:
# Print information about homelessness
hs.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 51 entries, 0 to 50
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   region          51 non-null     object 
 1   state           51 non-null     object 
 2   individuals     51 non-null     float64
 3   family_members  51 non-null     float64
 4   state_pop       51 non-null     int64  
dtypes: float64(2), int64(1), object(2)
memory usage: 2.4+ KB


In [ ]:
#Print the whole data
hs

region                 state  individuals  family_members  \
0   East South Central               Alabama       2570.0           864.0   
1              Pacific                Alaska       1434.0           582.0   
2             Mountain               Arizona       7259.0          2606.0   
3   West South Central              Arkansas       2280.0           432.0   
4              Pacific            California     109008.0         20964.0   
5             Mountain              Colorado       7607.0          3250.0   
6          New England           Connecticut       2280.0          1696.0   
7       South Atlantic              Delaware        708.0           374.0   
8       South Atlantic  District of Columbia       3770.0          3134.0   
9       South Atlantic               Florida      21443.0          9587.0   
10      South Atlantic               Georgia       6943.0          2556.0   
11             Pacific                Hawaii       4131.0          2399.0   
12            Mountain                 Idaho       1297.0           715.0   
13  East North Central              Illinois       6752.0          3891.0   
14  East North Central               Indiana       3776.0          1482.0   
15  West North Central                  Iowa       1711.0          1038.0   
16  West North Central                Kansas       1443.0           773.0   
17  East South Central              Kentucky       2735.0           953.0   
18  West South Central             Louisiana       2540.0           519.0   
19         New England                 Maine       1450.0          1066.0   
20      South Atlantic              Maryland       4914.0          2230.0   
21         New England         Massachusetts       6811.0         13257.0   
22  East North Central              Michigan       5209.0          3142.0   
23  West North Central             Minnesota       3993.0          3250.0   
24  East South Central           Mississippi       1024.0           328.0   
25  West North Central              Missouri       3776.0          2107.0   
26            Mountain               Montana        983.0           422.0   
27  West North Central              Nebraska       1745.0           676.0   
28            Mountain                Nevada       7058.0           486.0   
29         New England         New Hampshire        835.0           615.0   
30        Mid-Atlantic            New Jersey       6048.0          3350.0   
31            Mountain            New Mexico       1949.0           602.0   
32        Mid-Atlantic              New York      39827.0         52070.0   
33      South Atlantic        North Carolina       6451.0          2817.0   
34  West North Central          North Dakota        467.0            75.0   
35  East North Central                  Ohio       6929.0          3320.0   
36  West South Central              Oklahoma       2823.0          1048.0   
37             Pacific                Oregon      11139.0          3337.0   
38        Mid-Atlantic          Pennsylvania       8163.0          5349.0   
39         New England          Rhode Island        747.0           354.0   
40      South Atlantic        South Carolina       3082.0           851.0   
41  West North Central          South Dakota        836.0           323.0   
42  East South Central             Tennessee       6139.0          1744.0   
43  West South Central                 Texas      19199.0          6111.0   
44            Mountain                  Utah       1904.0           972.0   
45         New England               Vermont        780.0           511.0   
46      South Atlantic              Virginia       3928.0          2047.0   
47             Pacific            Washington      16424.0          5880.0   
48      South Atlantic         West Virginia       1021.0           222.0   
49  East North Central             Wisconsin       2740.0          2167.0   
50            Mountain               Wyoming        434.0           205.0   

    state_p

In [ ]:
#Print the last row of the data
hs.tail(1)

region    state  individuals  family_members  state_pop
50  Mountain  Wyoming        434.0           205.0     577601

In [ ]:
#Get random 5 rows
hs.sample(5)

region      state  individuals  family_members  state_pop
20      South Atlantic   Maryland       4914.0          2230.0    6035802
23  West North Central  Minnesota       3993.0          3250.0    5606249
46      South Atlantic   Virginia       3928.0          2047.0    8501286
2             Mountain    Arizona       7259.0          2606.0    7158024
44            Mountain       Utah       1904.0           972.0    3153550

In [ ]:
#in order to access specific column,
hs['region']

0     East South Central
1                Pacific
2               Mountain
3     West South Central
4                Pacific
5               Mountain
6            New England
7         South Atlantic
8         South Atlantic
9         South Atlantic
10        South Atlantic
11               Pacific
12              Mountain
13    East North Central
14    East North Central
15    West North Central
16    West North Central
17    East South Central
18    West South Central
19           New England
20        South Atlantic
21           New England
22    East North Central
23    West North Central
24    East South Central
25    West North Central
26              Mountain
27    West North Central
28              Mountain
29           New England
30          Mid-Atlantic
31              Mountain
32          Mid-Atlantic
33        South Atlantic
34    West North Central
35    East North Central
36    West South Central
37               Pacific
38          Mid-Atlantic
39           New England


In [ ]:
#or you can put . and column name after dataframe name
hs.region


<li>Print the number of rows and columns in <code>homelessness</code>.</li>


In [ ]:
# Print the shape of homelessness
hs.shape


(51, 5)

In [ ]:
#get the rows and columns
print(hs.shape[0])
print(hs.shape[1])

51
5


In [ ]:
#Print the lenght of homelessness data
len(hs)

51

In [ ]:
len(hs.columns)

5


<li>Print some summary statistics that describe the <code>homelessness</code> DataFrame.</li>


In [ ]:
# Print a summary stats of homelesness data
hs.describe()

individuals  family_members     state_pop
count      51.000000       51.000000  5.100000e+01
mean     7225.784314     3504.882353  6.405637e+06
std     15991.025083     7805.411811  7.327258e+06
min       434.000000       75.000000  5.776010e+05
25%      1446.500000      592.000000  1.777414e+06
50%      3082.000000     1482.000000  4.461153e+06
75%      6781.500000     3196.000000  7.340946e+06
max    109008.000000    52070.000000  3.946159e+07

<p class="">Insightful inspecting! You can see that the average number of homeless individuals in each state is about 7226. Let's explore the DataFrame further.</p>



In [ ]:
#In order to see statistics on non-numerical features,
#one has to explicitly indicate data types of interest in the include parameter.
hs.describe(include=['object'])

region    state
count               51       51
unique               9       51
top     South Atlantic  Alabama
freq                 9        1

In [ ]:
#Get frequincies by region-value_counts()
hs['region'].value_counts()

South Atlantic        9
Mountain              8
West North Central    7
New England           6
Pacific               5
East North Central    5
East South Central    4
West South Central    4
Mid-Atlantic          3
Name: region, dtype: int64

In [ ]:
#normalize=True is for %
hs['region'].value_counts(normalize=True)

South Atlantic        0.176471
Mountain              0.156863
West North Central    0.137255
New England           0.117647
Pacific               0.098039
East North Central    0.098039
East South Central    0.078431
West South Central    0.078431
Mid-Atlantic          0.058824
Name: region, dtype: float64

In [ ]:
#Get the correlation matrix of homeleness dataframe
hs.corr(numeric_only=True)

individuals  family_members  state_pop
individuals        1.000000        0.631426   0.834960
family_members     0.631426        1.000000   0.589913
state_pop          0.834960        0.589913   1.000000

In [ ]:
hs[['individuals', 'family_members']].corr()

individuals  family_members
individuals        1.000000        0.631426
family_members     0.631426        1.000000

In [ ]:
hs['individuals'].corr(hs['family_members'])

0.6314256443981799

### Parts of a DataFrame


<div class>
<p>To better understand DataFrame objects, it's useful to know that they consist of three components, stored as attributes:</p>
<ul>
<li>
<code>.values</code>: A two-dimensional NumPy array of values.</li>
<li>
<code>.columns</code>: An index of columns: the column names.</li>
<li>
<code>.index</code>: An index for the rows: either row numbers or row names.</li>
</ul>
<p>You can usually think of indexes as a list of strings or numbers, though the pandas <code>Index</code> data type allows for more sophisticated options. (These will be covered later in the course.)</p>
<p><code>homelessness</code> is available.</p>
</div>

<li>Import <code>pandas</code> using the alias <code>pd</code>.</li>


In [ ]:
# Import pandas using the alias pd
import pandas as pd


<li>Print a 2D NumPy array of the values in <code>homelessness</code>.</li>


In [ ]:
# Print the values of homelessness

hs.values

array([['East South Central', 'Alabama', 2570.0, 864.0, 4887681],
       ['Pacific', 'Alaska', 1434.0, 582.0, 735139],
       ['Mountain', 'Arizona', 7259.0, 2606.0, 7158024],
       ['West South Central', 'Arkansas', 2280.0, 432.0, 3009733],
       ['Pacific', 'California', 109008.0, 20964.0, 39461588],
       ['Mountain', 'Colorado', 7607.0, 3250.0, 5691287],
       ['New England', 'Connecticut', 2280.0, 1696.0, 3571520],
       ['South Atlantic', 'Delaware', 708.0, 374.0, 965479],
       ['South Atlantic', 'District of Columbia', 3770.0, 3134.0, 701547],
       ['South Atlantic', 'Florida', 21443.0, 9587.0, 21244317],
       ['South Atlantic', 'Georgia', 6943.0, 2556.0, 10511131],
       ['Pacific', 'Hawaii', 4131.0, 2399.0, 1420593],
       ['Mountain', 'Idaho', 1297.0, 715.0, 1750536],
       ['East North Central', 'Illinois', 6752.0, 3891.0, 12723071],
       ['East North Central', 'Indiana', 3776.0, 1482.0, 6695497],
       ['West North Central', 'Iowa', 1711.0, 1038.0, 3148618]

<li>Print the column names of <code>homelessness</code>.</li>


In [ ]:
# Print the column index of homelessness

hs.columns

Index(['region', 'state', 'individuals', 'family_members', 'state_pop'], dtype='object')

In [ ]:
for c in hs.columns:
  print(c)

region
state
individuals
family_members
state_pop


<li>Print the index of <code>homelessness</code>.</li>


In [ ]:
# Print the row index of homelessness
hs.index

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49,
            50],
           dtype='int64')

<p class="">Dynamite <code>DataFrame</code> dissection! <code>DataFrame</code>s have three components: values, a column index, and a row index.</p>



## Sorting and subsetting





### Sorting rows


<div class>
<p>Finding interesting bits of data in a DataFrame is often easier if you change the order of the rows. You can sort the rows by passing a column name to <code>.sort_values()</code>.</p>
<p>In cases where rows have the same value (this is common if you sort on a categorical variable), you may wish to break the ties by sorting on another column. You can sort on multiple columns in this way by passing a list of column names.</p>
<table>
<thead><tr>
<th>Sort on …</th>
<th>Syntax</th>
</tr></thead>
<tbody>
<tr>
<td>one column</td>
<td><code>df.sort_values("breed")</code></td>
</tr>
<tr>
<td>multiple columns</td>
<td><code>df.sort_values(["breed", "weight_kg"])</code></td>
</tr>
</tbody>
</table>
<p>By combining <code>.sort_values()</code> with <code>.head()</code>, you can answer questions in the form, "What are the top cases where…?".</p>
<p><code>homelessness</code> is available and <code>pandas</code> is loaded as <code>pd</code>.</p>
</div>

<li>Sort <code>homelessness</code> by the number of homeless individuals, from smallest to largest, and save this as <code>homelessness_ind</code>.</li>


In [ ]:
# Sort homelessness by individuals and print the top few rows
hs.sort_values("individuals").head()

region         state  individuals  family_members  state_pop
50            Mountain       Wyoming        434.0           205.0     577601
34  West North Central  North Dakota        467.0            75.0     758080
7       South Atlantic      Delaware        708.0           374.0     965479
39         New England  Rhode Island        747.0           354.0    1058287
45         New England       Vermont        780.0           511.0     624358


<li>Sort <code>homelessness</code> by the number of homeless <code>family_members</code> in descending order, and save this as <code>homelessness_fam</code>.</li>


In [ ]:
# Sort homelessness by descending family members and print the top few rows

hs.sort_values("family_members",ascending=False).head()

region          state  individuals  family_members  state_pop
32        Mid-Atlantic       New York      39827.0         52070.0   19530351
4              Pacific     California     109008.0         20964.0   39461588
21         New England  Massachusetts       6811.0         13257.0    6882635
9       South Atlantic        Florida      21443.0          9587.0   21244317
43  West South Central          Texas      19199.0          6111.0   28628666


<li>Sort <code>homelessness</code> first by region (ascending), and then by number of family members (descending). Save this as <code>homelessness_reg_fam</code>.</li>


In [ ]:
#  Sort homelessness by region, then descending family members and Print the top few rows
hs.sort_values(['region',"family_members"],ascending=[True,False]).head()

region      state  individuals  family_members  state_pop
13  East North Central   Illinois       6752.0          3891.0   12723071
35  East North Central       Ohio       6929.0          3320.0   11676341
22  East North Central   Michigan       5209.0          3142.0    9984072
49  East North Central  Wisconsin       2740.0          2167.0    5807406
14  East North Central    Indiana       3776.0          1482.0    6695497

<p class="">Your head is sorted! Using <code>.sort_values()</code> and <code>.head()</code> form a powerful pair.</p>



### Subsetting columns


<div class>
<p>When working with data, you may not need all of the variables in your dataset. Square brackets (<code>[]</code>) can be used to select only the columns that matter to you in an order that makes sense to you.
To select only <code>"col_a"</code> of the DataFrame <code>df</code>, use</p>
<pre><code>df["col_a"]
</code></pre>
<p>To select <code>"col_a"</code> and <code>"col_b"</code> of <code>df</code>, use</p>
<pre><code>df[["col_a", "col_b"]]
</code></pre>
<p><code>homelessness</code> is available and <code>pandas</code> is loaded as <code>pd</code>.</p>
</div>

<li>Create a DataFrame called <code>individuals</code> that contains only the <code>individuals</code> column of <code>homelessness</code>.</li>


In [ ]:
# Select the individuals column Print the head of the result
hs['individuals'].head()

0      2570.0
1      1434.0
2      7259.0
3      2280.0
4    109008.0
Name: individuals, dtype: float64


<li>Create a DataFrame called <code>state_fam</code> that contains only the <code>state</code> and <code>family_members</code> columns of <code>homelessness</code>, in that order.</li>


In [ ]:
# Select the state and family_members columns and print the head of the result
hs[['state' , 'family_members']].head()


state  family_members
0     Alabama           864.0
1      Alaska           582.0
2     Arizona          2606.0
3    Arkansas           432.0
4  California         20964.0


<li>Create a DataFrame called <code>ind_state</code> that contains the <code>individuals</code> and <code>state</code> columns of <code>homelessness</code>, in that order.</li>


In [ ]:
# Select only the individuals and state columns, in that order Print the head of the result
ind_state= hs[['individuals','state']].head()
ind_state.head()


individuals       state
0       2570.0     Alabama
1       1434.0      Alaska
2       7259.0     Arizona
3       2280.0    Arkansas
4     109008.0  California

<p class="">Radical reordering! Selecting and reordering columns can make data easier to work with.</p>



### Subsetting rows


<div class>
<p>A large part of data science is about finding which bits of your dataset are interesting. One of the simplest techniques for this is to find a subset of rows that match some criteria. This is sometimes known as <em>filtering rows</em> or <em>selecting rows</em>.</p>
<p>There are many ways to subset a DataFrame, perhaps the most common is to use relational operators to return <code>True</code> or <code>False</code> for each row, then pass that inside square brackets.</p>
<pre><code>dogs[dogs["height_cm"] &gt; 60]
dogs[dogs["color"] == "tan"]
</code></pre>
<p>You can filter for multiple conditions at once by using the "bitwise and" operator, <code>&amp;</code>.</p>
<pre><code>dogs[(dogs["height_cm"] &gt; 60) &amp; (dogs["color"] == "tan")]
</code></pre>
<p><code>homelessness</code> is available and <code>pandas</code> is loaded as <code>pd</code>.</p>
</div>
<div class="exercise--instructions__content"><p>Filter <code>homelessness</code> for cases where the number of individuals is greater than ten thousand, assigning to <code>ind_gt_10k</code>. <em>View the printed result.</em></p></div>


In [ ]:
# Filter for rows where individuals is greater than 10000

ind_gt_10k=hs[hs['individuals']>10000]
# See the result
ind_gt_10k


region       state  individuals  family_members  state_pop
4              Pacific  California     109008.0         20964.0   39461588
9       South Atlantic     Florida      21443.0          9587.0   21244317
32        Mid-Atlantic    New York      39827.0         52070.0   19530351
37             Pacific      Oregon      11139.0          3337.0    4181886
43  West South Central       Texas      19199.0          6111.0   28628666
47             Pacific  Washington      16424.0          5880.0    7523869


<div class="exercise--instructions__content"><p>Filter <code>homelessness</code> for cases where the USA Census region is <code>"Mountain"</code>, assigning to <code>mountain_reg</code>. <em>View the printed result.</em></p></div>


In [ ]:
# Filter for rows where region is Mountain
mountain_reg=hs[hs['region']=="Mountain"]
# See the result
print(mountain_reg)


      region       state  individuals  family_members  state_pop
2   Mountain     Arizona       7259.0          2606.0    7158024
5   Mountain    Colorado       7607.0          3250.0    5691287
12  Mountain       Idaho       1297.0           715.0    1750536
26  Mountain     Montana        983.0           422.0    1060665
28  Mountain      Nevada       7058.0           486.0    3027341
31  Mountain  New Mexico       1949.0           602.0    2092741
44  Mountain        Utah       1904.0           972.0    3153550
50  Mountain     Wyoming        434.0           205.0     577601



<div class="exercise--instructions__content"><p>Filter <code>homelessness</code> for cases where the number of <code>family_members</code> is less than one thousand and the <code>region</code> is "Pacific", assigning to <code>fam_lt_1k_pac</code>. <em>View the printed result.</em></p></div>


In [ ]:
# Filter for rows where family_members is less than 1000
# and region is Pacific
fam_lt_1k_pac=hs[(hs['family_members']<1000) & (hs['region']=='Pacific')
# See the result
print(fam_lt_1k_pac)


    region   state  individuals  family_members  state_pop
1  Pacific  Alaska       1434.0           582.0     735139


<p class="">Superb subsetting! Using square brackets plus logical conditions is often the most powerful way of identifying interesting rows of data.</p>



### Subsetting rows by categorical variables


<div class>
<p>Subsetting data based on a categorical variable often involves using the "or" operator (<code>|</code>) to select rows from multiple categories. This can get tedious when you want all states in one of three different regions, for example.
Instead, use the <code>.isin()</code> method, which will allow you to tackle this problem by writing one condition instead of three separate ones.</p>
<pre><code>colors = ["brown", "black", "tan"]
condition = dogs["color"].isin(colors)
dogs[condition]
</code></pre>
<p><code>homelessness</code> is available and <code>pandas</code> is loaded as <code>pd</code>.</p>
</div>
<div class="exercise--instructions__content"><p>Filter <code>homelessness</code> for cases where the USA census region is "South Atlantic" or it is "Mid-Atlantic", assigning to <code>south_mid_atlantic</code>. <em>View the printed result.</em></p></div>


In [ ]:
# Subset for rows in South Atlantic or Mid-Atlantic regions

# See the result
south_mid_atlantic=hs[(hs['region']=='South Atlantic') | (hs['region']=='Mid-Atlantic')]
south_mid_atlantic


region                 state  individuals  family_members  \
7   South Atlantic              Delaware        708.0           374.0   
8   South Atlantic  District of Columbia       3770.0          3134.0   
9   South Atlantic               Florida      21443.0          9587.0   
10  South Atlantic               Georgia       6943.0          2556.0   
20  South Atlantic              Maryland       4914.0          2230.0   
30    Mid-Atlantic            New Jersey       6048.0          3350.0   
32    Mid-Atlantic              New York      39827.0         52070.0   
33  South Atlantic        North Carolina       6451.0          2817.0   
38    Mid-Atlantic          Pennsylvania       8163.0          5349.0   
40  South Atlantic        South Carolina       3082.0           851.0   
46  South Atlantic              Virginia       3928.0          2047.0   
48  South Atlantic         West Virginia       1021.0           222.0   

    state_pop  
7      965479  
8      701547  
9    21244317  
10   10511131  
20    6035802  
30    8886025  
32   19530351  
33   10381615  
38   12800922  
40    5084156  
46    8501286  
48    1804291

In [ ]:
reg=['South Atlantic','Mid-Atlantic']
condition = hs["region"].isin(reg)
hs[condition]

region                 state  individuals  family_members  \
7   South Atlantic              Delaware        708.0           374.0   
8   South Atlantic  District of Columbia       3770.0          3134.0   
9   South Atlantic               Florida      21443.0          9587.0   
10  South Atlantic               Georgia       6943.0          2556.0   
20  South Atlantic              Maryland       4914.0          2230.0   
30    Mid-Atlantic            New Jersey       6048.0          3350.0   
32    Mid-Atlantic              New York      39827.0         52070.0   
33  South Atlantic        North Carolina       6451.0          2817.0   
38    Mid-Atlantic          Pennsylvania       8163.0          5349.0   
40  South Atlantic        South Carolina       3082.0           851.0   
46  South Atlantic              Virginia       3928.0          2047.0   
48  South Atlantic         West Virginia       1021.0           222.0   

    state_pop  
7      965479  
8      701547  
9    21244317  
10   10511131  
20    6035802  
30    8886025  
32   19530351  
33   10381615  
38   12800922  
40    5084156  
46    8501286  
48    1804291


<div class="exercise--instructions__content"><p>Filter <code>homelessness</code> for cases where the USA census <code>state</code> is in the list of Mojave states, <code>canu</code>, assigning to <code>mojave_homelessness</code>. <em>View the printed result.</em></p></div>


In [ ]:
# The Mojave Desert states
canu = ["California", "Arizona", "Nevada", "Utah"]

# Filter for rows in the Mojave Desert states using isin()
mojave_hs = hs["state"].isin(canu)

# See the result
print(hs[mojave_hs])


      region       state  individuals  family_members  state_pop
2   Mountain     Arizona       7259.0          2606.0    7158024
4    Pacific  California     109008.0         20964.0   39461588
28  Mountain      Nevada       7058.0           486.0    3027341
44  Mountain        Utah       1904.0           972.0    3153550


<p class="">Success <em>is in</em> the air! Using <code>.isin()</code> makes subsetting categorical variables a breeze.</p>



## New columns





### Adding new columns


<div class>
<p>You aren't stuck with just the data you are given. Instead, you can add new columns to a DataFrame. This has many names, such as <em>transforming</em>, <em>mutating</em>, and <em>feature engineering</em>.</p>
<p>You can create new columns from scratch, but it is also common to derive them from other columns, for example, by adding columns together or by changing their units. </p>
<p><code>homelessness</code> is available and <code>pandas</code> is loaded as <code>pd</code>.</p>
</div>

<li>Add a new column to <code>homelessness</code>, named <code>total</code>, containing the sum of the <code>individuals</code> and <code>family_members</code> columns.</li>


In [ ]:
# Add total col as sum of individuals and family_members
hs["total"]=hs['individuals']+hs['family_members']
hs.head()

region       state  individuals  family_members  state_pop  \
0  East South Central     Alabama       2570.0           864.0    4887681   
1             Pacific      Alaska       1434.0           582.0     735139   
2            Mountain     Arizona       7259.0          2606.0    7158024   
3  West South Central    Arkansas       2280.0           432.0    3009733   
4             Pacific  California     109008.0         20964.0   39461588   

      total  
0    3434.0  
1    2016.0  
2    9865.0  
3    2712.0  
4  129972.0

<li>Add another column to <code>homelessness</code>, named <code>p_individuals</code>, containing the proportion of homeless people in each state who are individuals.</li>


In [ ]:
# Add p_individuals col as proportion of total that are individuals
hs["p_individuals"] = hs['individuals']/hs['total']

# See the result
hs.head()


region       state  individuals  family_members  state_pop  \
0  East South Central     Alabama       2570.0           864.0    4887681   
1             Pacific      Alaska       1434.0           582.0     735139   
2            Mountain     Arizona       7259.0          2606.0    7158024   
3  West South Central    Arkansas       2280.0           432.0    3009733   
4             Pacific  California     109008.0         20964.0   39461588   

      total  p_individuals  
0    3434.0       0.748398  
1    2016.0       0.711310  
2    9865.0       0.735834  
3    2712.0       0.840708  
4  129972.0       0.838704

<p class="">Astounding column assignment! If your dataset doesn't have the exact columns you need, you can often make your own from what you have.</p>



### Combo-attack!


<div class>
<p>You've seen the four most common types of data manipulation: sorting rows, subsetting columns, subsetting rows, and adding new columns. In a real-life data analysis, you can mix and match these four manipulations to answer a multitude of questions.</p>
<p>In this exercise, you'll answer the question, "Which state has the highest number of homeless individuals per 10,000 people in the state?" Combine your new <code>pandas</code> skills to find out.</p>
</div>

<li>Add a column to <code>homelessness</code>, <code>indiv_per_10k</code>, containing the number of homeless individuals per ten thousand people in each state.</li>


In [ ]:
# Create indiv_per_10k col as homeless individuals per 10k state pop
hs["indiv_per_10k"] =10000*hs['individuals']/hs['state_pop']


In [ ]:
hs.head()

region       state  individuals  family_members  state_pop  \
0  East South Central     Alabama       2570.0           864.0    4887681   
1             Pacific      Alaska       1434.0           582.0     735139   
2            Mountain     Arizona       7259.0          2606.0    7158024   
3  West South Central    Arkansas       2280.0           432.0    3009733   
4             Pacific  California     109008.0         20964.0   39461588   

      total  p_individuals  indiv_per_10k  
0    3434.0       0.748398       5.258117  
1    2016.0       0.711310      19.506515  
2    9865.0       0.735834      10.141067  
3    2712.0       0.840708       7.575423  
4  129972.0       0.838704      27.623825

<li>Subset rows where <code>indiv_per_10k</code> is higher than <code>20</code>, assigning to <code>high_homelessness</code>.</li>


In [ ]:
# Subset rows for indiv_per_10k greater than 20
high_hs = hs[hs['indiv_per_10k']>20]
high_hs


region                 state  individuals  family_members  \
4          Pacific            California     109008.0         20964.0   
8   South Atlantic  District of Columbia       3770.0          3134.0   
11         Pacific                Hawaii       4131.0          2399.0   
28        Mountain                Nevada       7058.0           486.0   
32    Mid-Atlantic              New York      39827.0         52070.0   
37         Pacific                Oregon      11139.0          3337.0   
47         Pacific            Washington      16424.0          5880.0   

    state_pop     total  p_individuals  indiv_per_10k  
4    39461588  129972.0       0.838704      27.623825  
8      701547    6904.0       0.546060      53.738381  
11    1420593    6530.0       0.632619      29.079406  
28    3027341    7544.0       0.935578      23.314189  
32   19530351   91897.0       0.433387      20.392363  
37    4181886   14476.0       0.769481      26.636307  
47    7523869   22304.0       0.736370      21.829195

<li>Sort <code>high_homelessness</code> by descending <code>indiv_per_10k</code>, assigning to <code>high_homelessness_srt</code>.</li>


In [ ]:
# Sort high_homelessness by descending indiv_per_10k
high_hs_srt = high_hs.sort_values(['indiv_per_10k'],ascending=[False])
high_hs_srt

region                 state  individuals  family_members  \
8   South Atlantic  District of Columbia       3770.0          3134.0   
11         Pacific                Hawaii       4131.0          2399.0   
4          Pacific            California     109008.0         20964.0   
37         Pacific                Oregon      11139.0          3337.0   
28        Mountain                Nevada       7058.0           486.0   
47         Pacific            Washington      16424.0          5880.0   
32    Mid-Atlantic              New York      39827.0         52070.0   

    state_pop     total  p_individuals  indiv_per_10k  
8      701547    6904.0       0.546060      53.738381  
11    1420593    6530.0       0.632619      29.079406  
4    39461588  129972.0       0.838704      27.623825  
37    4181886   14476.0       0.769481      26.636307  
28    3027341    7544.0       0.935578      23.314189  
47    7523869   22304.0       0.736370      21.829195  
32   19530351   91897.0       0.433387      20.392363

<li>Select only the <code>state</code> and <code>indiv_per_10k</code> columns of <code>high_homelessness_srt</code> and save as <code>result</code>. <em>Look at the <code>result</code>.</em>
</li>


In [ ]:
# From high_homelessness_srt, select the state and indiv_per_10k cols
result = high_hs_srt[['state',	'indiv_per_10k']]

# See the result
result


state  indiv_per_10k
8   District of Columbia      53.738381
11                Hawaii      29.079406
4             California      27.623825
37                Oregon      26.636307
28                Nevada      23.314189
47            Washington      21.829195
32              New York      20.392363

In [ ]:
#Write this as csv using to_csv
result.to_excel('/content/result.xlsx')

<p class="">Cool combination! District of Columbia has the highest number of homeless individuals - almost 54 per ten thousand people. This is almost double the number of the next-highest state, Hawaii. If you combine new column addition, row subsetting, sorting, and column selection, you can answer lots of questions like this.</p>



# Aggregating DataFrames

<p class="chapter__description">
    In this chapter, you’ll calculate summary statistics on DataFrame columns, and master grouped summary statistics and pivot tables.
  </p>



## Summary statistics





### Mean and median


<div class>
<p>Summary statistics are exactly what they sound like - they summarize many numbers in one statistic. For example, mean, median, minimum, maximum, and standard deviation are summary statistics. Calculating summary statistics allows you to get a better sense of your data, even if there's a lot of it. </p>

</div>


In [ ]:
#get the mean value of state pouplation
hs['state_pop'].mean()

6405637.274509804

<li>Print the median of the <code>weekly_sales</code> column.</li>


In [ ]:
#get the median value of state population
hs['state_pop'].median()

4461153.0

<p class="">The mean weekly sales amount is almost double the median weekly sales amount! This can tell you that there are a few very high sales weeks that are making the mean so much higher than the median.</p>



In [ ]:
#get the maximum value of state population
hs['state_pop'].max()

39461588

<li>Print the minimum of the <code>date</code> column.</li>


In [ ]:
#get the minimum value of state population

hs['state_pop'].min()

577601

<p class="">Super summarizing! Taking the minimum and maximum of a column of dates is handy for figuring out what time period your data covers. In this case, there are data from February of 2010 to October of 2012.</p>



### Calculations with .groupby()


<div class>
<p>The <code>.groupby()</code> method makes life much easier. In this exercise, you'll perform the same calculations as last time, except you'll use the <code>.groupby()</code> method.

In [ ]:
#Get region population
#Group by type; calculate state population by region
#stat_pop_by_region=hs.groupby('region').sum()['state_pop']
stat_pop_by_region=hs.groupby('region').sum()['state_pop']
stat_pop_by_region

region
East North Central    46886387
East South Central    19101485
Mid-Atlantic          41217298
Mountain              24511745
New England           14829322
Pacific               53323075
South Atlantic        65229624
West North Central    21350241
West South Central    40238324
Name: state_pop, dtype: int64

### Multiple grouped summaries


<div class>
<p>Earlier in this chapter, you saw that the <code>.agg()</code> method is useful to compute multiple statistics on multiple variables. It also works with grouped data. NumPy, which is imported as <code>np</code>, has many different summary statistics functions, including: <code>np.min</code>, <code>np.max</code>, <code>np.mean</code>, and <code>np.median</code>.</p>

</div>

<li>Import <code>numpy</code> with the alias <code>np</code>.</li>


In [ ]:
#Get all statistics for region population
#df.groupby(["Churn"])[columns_to_show].agg([np.mean, np.std, np.min, np.max])
import numpy as np
stat_pop_by_region_all=hs.groupby('region')['state_pop'].agg([np.mean, np.std, np.min, np.max, np.sum])
stat_pop_by_region_all

mean           std     amin      amax       sum
region                                                                     
East North Central  9.377277e+06  3.032497e+06  5807406  12723071  46886387
East South Central  4.775371e+06  1.561633e+06  2981020   6771631  19101485
Mid-Atlantic        1.373910e+07  5.383823e+06  8886025  19530351  41217298
Mountain            3.063968e+06  2.284641e+06   577601   7158024  24511745
New England         2.471554e+06  2.392057e+06   624358   6882635  14829322
Pacific             1.066462e+07  1.631872e+07   735139  39461588  53323075
South Atlantic      7.247736e+06  6.481193e+06   701547  21244317  65229624
West North Central  3.050034e+06  2.130333e+06   758080   6121623  21350241
West South Central  1.005958e+07  1.239780e+07  3009733  28628666  40238324

<p class="">Awesome aggregating! Notice that the minimum <code>weekly_sales</code> is negative because some stores had more returns than sales.</p>



## `SQL`

In [ ]:
#Install Sql library fo pandas
!pip install pandasql
!pip install SQLAlchemy==1.4.46

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
#You can get this using sql as well usng pandasql


import pandasql as psql
qr= psql.sqldf('select region, min(state_pop),max(state_pop),sum(state_pop) ,avg(state_pop) from hs group by region')
qr

region  min(state_pop)  max(state_pop)  sum(state_pop)  \
0  East North Central         5807406        12723071        46886387   
1  East South Central         2981020         6771631        19101485   
2        Mid-Atlantic         8886025        19530351        41217298   
3            Mountain          577601         7158024        24511745   
4         New England          624358         6882635        14829322   
5             Pacific          735139        39461588        53323075   
6      South Atlantic          701547        21244317        65229624   
7  West North Central          758080         6121623        21350241   
8  West South Central         3009733        28628666        40238324   

   avg(state_pop)  
0    9.377277e+06  
1    4.775371e+06  
2    1.373910e+07  
3    3.063968e+06  
4    2.471554e+06  
5    1.066462e+07  
6    7.247736e+06  
7    3.050034e+06  
8    1.005958e+07

In [ ]:
hs

region                 state  individuals  family_members  \
0   East South Central               Alabama       2570.0           864.0   
1              Pacific                Alaska       1434.0           582.0   
2             Mountain               Arizona       7259.0          2606.0   
3   West South Central              Arkansas       2280.0           432.0   
4              Pacific            California     109008.0         20964.0   
5             Mountain              Colorado       7607.0          3250.0   
6          New England           Connecticut       2280.0          1696.0   
7       South Atlantic              Delaware        708.0           374.0   
8       South Atlantic  District of Columbia       3770.0          3134.0   
9       South Atlantic               Florida      21443.0          9587.0   
10      South Atlantic               Georgia       6943.0          2556.0   
11             Pacific                Hawaii       4131.0          2399.0   
12            Mountain                 Idaho       1297.0           715.0   
13  East North Central              Illinois       6752.0          3891.0   
14  East North Central               Indiana       3776.0          1482.0   
15  West North Central                  Iowa       1711.0          1038.0   
16  West North Central                Kansas       1443.0           773.0   
17  East South Central              Kentucky       2735.0           953.0   
18  West South Central             Louisiana       2540.0           519.0   
19         New England                 Maine       1450.0          1066.0   
20      South Atlantic              Maryland       4914.0          2230.0   
21         New England         Massachusetts       6811.0         13257.0   
22  East North Central              Michigan       5209.0          3142.0   
23  West North Central             Minnesota       3993.0          3250.0   
24  East South Central           Mississippi       1024.0           328.0   
25  West North Central              Missouri       3776.0          2107.0   
26            Mountain               Montana        983.0           422.0   
27  West North Central              Nebraska       1745.0           676.0   
28            Mountain                Nevada       7058.0           486.0   
29         New England         New Hampshire        835.0           615.0   
30        Mid-Atlantic            New Jersey       6048.0          3350.0   
31            Mountain            New Mexico       1949.0           602.0   
32        Mid-Atlantic              New York      39827.0         52070.0   
33      South Atlantic        North Carolina       6451.0          2817.0   
34  West North Central          North Dakota        467.0            75.0   
35  East North Central                  Ohio       6929.0          3320.0   
36  West South Central              Oklahoma       2823.0          1048.0   
37             Pacific                Oregon      11139.0          3337.0   
38        Mid-Atlantic          Pennsylvania       8163.0          5349.0   
39         New England          Rhode Island        747.0           354.0   
40      South Atlantic        South Carolina       3082.0           851.0   
41  West North Central          South Dakota        836.0           323.0   
42  East South Central             Tennessee       6139.0          1744.0   
43  West South Central                 Texas      19199.0          6111.0   
44            Mountain                  Utah       1904.0           972.0   
45         New England               Vermont        780.0           511.0   
46      South Atlantic              Virginia       3928.0          2047.0   
47             Pacific            Washington      16424.0          5880.0   
48      South Atlantic         West Virginia       1021.0           222.0   
49  East North Central             Wisconsin       2740.0          2167.0   
50            Mountain               Wyoming        434.0           205.0   

    state_p

If else

In [ ]:
#mean of state-pop
hs['state_pop'].mean()

6405637.274509804

In [ ]:
#Create new variable nmaed 'state_pop_LH' and classify state pop hig or low based on mean value of state population
import numpy as np
hs['state_pop_LH']=np.where(hs['state_pop']>hs['state_pop'].mean(),'High','Low')
hs.head()

region       state  individuals  family_members  state_pop  \
0  East South Central     Alabama       2570.0           864.0    4887681   
1             Pacific      Alaska       1434.0           582.0     735139   
2            Mountain     Arizona       7259.0          2606.0    7158024   
3  West South Central    Arkansas       2280.0           432.0    3009733   
4             Pacific  California     109008.0         20964.0   39461588   

      total  p_individuals  indiv_per_10k state_pop_LH  
0    3434.0       0.748398       5.258117          Low  
1    2016.0       0.711310      19.506515          Low  
2    9865.0       0.735834      10.141067         High  
3    2712.0       0.840708       7.575423          Low  
4  129972.0       0.838704      27.623825         High

In [ ]:
#drop newly created variable 'state_pop_LH'.
#df = df.drop('column_name', axis=1)
#hs = df.drop('column_name', axis=1)
hs=hs.drop(['total',	'p_individuals',	'indiv_per_10k'	,'state_pop_LH'	,'state_pop_LH2'], axis=1)
hs.head()

region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588